### ISSUE TITLE
 LinAlgError: Singular matrix by RegressionCorrector

### ISSUE THREAD

**Issue by ansainbdg** [KeplerGO/lightkurve#648]
LinAlgError: Singular matrix by RegressionCorrecor

I tried to do the tutorial
”https://docs.lightkurve.org/tutorials/04-how-to-remove-tess-scattered-light-using-regressioncorrector.html　
2.Using RegressionCorrector on TESS 2-minute cadence Target Pixel Files” 
on another star.

However, the program did not work as per the tutorial and gave the following error when omi02 Eri, HD50281,58Eri.

corrected_lc = reg . correct ( dm )
C:\Users\tit1c_000\Anaconda3\lib\site-packages\lightkurve\correctors\regressioncorrector.py:149: RuntimeWarning: divide by zero encountered in true_divide
  sigma_w_inv = np.dot(X.T, X / flux_err[:, None]**2)
C:\Users\tit1c_000\Anaconda3\lib\site-packages\lightkurve\correctors\regressioncorrector.py:154: RuntimeWarning: divide by zero encountered in true_divide
  B = np.dot(X.T, self.lc.flux[cadence_mask] / flux_err**2)
WARNING: Input data contains invalid values (NaNs or infs), which were automatically clipped. [astropy.stats.sigma_clipping]
---------------------------------------------------------------------------
LinAlgError                               Traceback (most recent call last)
<ipython-input-12-74f8a36ec14d> in <module>
----> 1 corrected_lc = reg . correct ( dm )

~\Anaconda3\lib\site-packages\lightkurve\correctors\regressioncorrector.py in correct(self, design_matrix_collection, cadence_mask, sigma, niters, propagate_errors)
    209                                        prior_mu=self.X.prior_mu,
    210                                        prior_sigma=self.X.prior_sigma,
--> 211                                        propagate_errors=propagate_errors)
    212             residuals = self.lc.flux - np.dot(self.X.values, coefficients)
    213             cadence_mask &= ~sigma_clip(residuals, sigma=sigma).mask

~\Anaconda3\lib\site-packages\lightkurve\correctors\regressioncorrector.py in _fit_coefficients(self, cadence_mask, prior_mu, prior_sigma, propagate_errors)
    157 
    158         # Solve for weights w
--> 159         w = np.linalg.solve(sigma_w_inv, B).T
    160         if propagate_errors:
    161             w_err = np.linalg.inv(sigma_w_inv)

<__array_function__ internals> in solve(*args, **kwargs)

~\Anaconda3\lib\site-packages\numpy\linalg\linalg.py in solve(a, b)
    401     signature = 'DD->D' if isComplexType(t) else 'dd->d'
    402     extobj = get_linalg_error_extobj(_raise_linalgerror_singular)
--> 403     r = gufunc(a, b, signature=signature, extobj=extobj)
    404 
    405     return wrap(r.astype(result_t, copy=False))

~\Anaconda3\lib\site-packages\numpy\linalg\linalg.py in _raise_linalgerror_singular(err, flag)
     95 
     96 def _raise_linalgerror_singular(err, flag):
---> 97     raise LinAlgError("Singular matrix")
     98 
     99 def _raise_linalgerror_nonposdef(err, flag):

LinAlgError: Singular matrix

#### Environment

-  platform (windows jupyter notebook):
-  lightkurve version (1.5.2):
-  installation method (pip):

**Comment by barentsen** [comment]
@ansainbdg Thank you for reporting this issue! :+1:

If possible, would you have the full code snippet I can use to reproduce the bug?

**Comment by barentsen** [comment]
(For reference: my suspicion is that this issue is caused because some FLUX_ERR values are NaN or zero.)

**Comment by ansainbdg** [comment]
import lightkurve as lk
import numpy as np
import matplotlib.pyplot as plt
target = 'HD50281'
tpf_2min = lk . search_targetpixelfile ( target , mission = 'TESS' , sector = 6 ) . download ()
# Use the pipeline aperture and a raw light curve
aper = tpf_2min . pipeline_mask
raw_lc = tpf_2min . to_lightcurve ()

# Make a design matrix
dm = lk . DesignMatrix ( tpf_2min . flux [:, ~ aper ], name = 'pixels' ) . pca ( 5 ) . append_constant ()

# Regression Corrector Object
reg = lk . RegressionCorrector ( raw_lc )
corrected_lc = reg . correct ( dm )

After this I get the above error.
The same error occurs even if the target and sector are changed to omi02 Eri. However, there is no error when changing to KIC 8462852(tutorial).

**Comment by barentsen** [comment]
@ansainbdg Thank you so much!  I hope to debug this issue first thing tomorrow morning.

**Comment by barentsen** [comment]
The cause of this issue is the fact that cadences which have the Argabrightening quality flag set (i.e. quality=16) contain NaN values in the FLUX_ERR and FLUX_BKG columns of the Target Pixel File products.

@ansainbdg The quick fix to the issue is to replace the following line in your notebook
```python
tpf_2min = lk.search_targetpixelfile(target, mission='TESS', sector=6).download()
```
with this:
```python
tpf_2min = lk.search_targetpixelfile(target, mission='TESS', sector=6).download(quality_bitmask="hardest")
```
(because it will ignore the problem cadences).

However it looks like we do need to make Lightkurve's RegressionCorrector more robust against the presence of NaN values!

**Comment by barentsen** [comment]
This issue should have been fixed by #651 and will be included in a soon-to-be-released bugfix release (v1.5.3).

**Comment by ansainbdg** [comment]
thanks!!
